### ISSUE TITLE
docker build --pull issue with docker-engine 1.13

### ISSUE THREAD

**Issue by gengmao** [awslabs/amazon-ecr-credential-helper#23]
Can easily test with a simple dockerfile
```
FROM <id>.dkr.ecr.us-east-1.amazonaws.com/ubuntu14.04:latest
CMD ["echo", "test"]
```
Docker 1.13, build it with --pull option will fail.
```
docker build --pull -t test .
Sending build context to Docker daemon 2.048 kB
Step 1/2 : FROM <id>.dkr.ecr.us-east-1.amazonaws.com/ubuntu14.04:latest
Pulling repository <id>.dkr.ecr.us-east-1.amazonaws.com/ubuntu14.04
unauthorized: authentication required
```
Errors in /var/log/upstart/docker.log
```
time="2017-02-08T17:49:47.089538193Z" level=error msg="Attempting next endpoint for pull after error: Get https://<id>.dkr.ecr.us-east-1.amazonaws.com/v2/ubuntu14.04/manifests/latest: no basic auth credentials"
time="2017-02-08T17:49:47.104752428Z" level=error msg="Not continuing with pull after error: unauthorized: authentication required"
```
.docker/config.json:
```
{
    "credsStore": "ecr-login",
    "auths": {
        "<id>.dkr.ecr.us-east-1.amazonaws.com": {
        }
    }
}
```
Tested several workarounds:
- docker 1.12 has no problem
- docker 1.13, removing --pull option, docker build is fine
- docker 1.13, removing "credsStore": "ecr-login" from config.json, use aws ecr get-login, docker build --pull is fine.

**Comment by samuelkarp** [comment]
It looks like Docker changed this in a backwards-incompatible way in Docker 1.13.  This credential helper will need to be updated to provide the new `client.List` API (see https://github.com/docker/docker-credential-helpers/pull/25 and https://github.com/docker/docker/pull/26354).

I don't have an estimate for when I'd be able to get around to this.  If you're interested in taking a stab at it, we should be able to reuse the [credential cache implementation](https://github.com/awslabs/amazon-ecr-credential-helper/blob/192a4928ba97539621c6f0cffcc21d6853bb5921/ecr-login/cache/file.go) to populate this list.  That would also change the workflow from specifying the registries in `~/.docker/config.json` to running a pull/push once in order to ensure that the registry is present in the cache file.

**Comment by gengmao** [comment]
@samuelkarp thanks for quick response. I will study the links a bit.

**Comment by jaredm4** [comment]
I'm pretty new to this credential helper, but could this backwards-incompatible change be what's causing my issue that I filed with Docker? https://github.com/docker/docker/issues/31063

**Comment by samuelkarp** [comment]
@jaredm4 I don't know enough about how Swarm-mode handles authentication to know whether it's related.  However, my suspicion would be that it's unrelated; if I remember correctly Swarm-mode does not handle credential rotation without restarting all the containers.

**Comment by sentientmonkey** [comment]
Started working on this a bit (at least vending and bringing in the List interface) in #26.

**Comment by gengmao** [comment]
@sentientmonkey Hi Scott! Glad to see you here :) Thanks for helping on this 👍  I took a look at the sources but was distracted by other things... 
One thing I was thinking is how to figure out a fix which is backwards-compatible with docker-engine < 1.13 and old config.json? If it is too difficult to achieve, that's understandable - I will plan to upgrade docker-engine, ecr-login, config.json together then.

**Comment by sentientmonkey** [comment]
:wave: :smile:
Reviewing over a few issues, I found this PR docker/docker#26354 - looks like they moved to the credential helper interface & added a dependency on List so they could fix build for credential helpers. docker/docker-credential-helpers#25 first implemented the List API, which wasn't in the original interface when we first implemented our helper. To keep it backwards compatible in 1.13 Docker would have had to retain the older code paths that fall back if List isn't implemented, but I don't think they fully considered other credential helpers since they ship the default credential helpers in Docker.

**Comment by gengmao** [comment]
@sentientmonkey oh I was asking the reversed backword-compatibility. When your fix added the List interface implementation to ecr-login helper, will it become incompatible with docker-engine 1.12 and old config.json (I thought it needs changes to make ecr-login helper work with docker 1.13, isn't it)?

**Comment by sentientmonkey** [comment]
@gengmao Looks like it's a matter of implementing the list command from the credential helper side. This might not help older versions of docker (i.e. 1.12), but for 1.13 and newer this should be a drop-in replacement that will fix build in many cases (if you've already pulled or pushed to that repo ahead of time). I'm not sure if adding the entry in the config.json will to continue to work after the list is implemented - will have to test and see.

**Comment by sentientmonkey** [comment]
@gengmao I've merged in the fix - this should allow list to work without having to add entries in the config.json. You will have to pull (or push) an image before referencing it in your build, but this should work slightly better than before on 1.13.

**Comment by gengmao** [comment]
Thanks a lot @sentientmonkey! I tested and looks it works with both docker 1.12 and docker 1.13. The config.json needs not to change too. 
I am interested in how ecr-login helper can support List without cache exist and how it can support the new "credHelpers" config like https://github.com/GoogleCloudPlatform/docker-credential-gcr, but that's not urgent.

**Comment by joshzarrabi** [comment]
Wait, so do I have it correct that the fix is in, but for it to work as intended, you need to have downloaded the image that's in your `FROM` first?

**Comment by samuelkarp** [comment]
> Wait, so do I have it correct that the fix is in, but for it to work as intended, you need to have downloaded the image that's in your FROM first?

No, you need to pull or push an image in the same registry (not repository) or use the registry associated with your account.

**Comment by chendrix** [comment]
@samuelkarp interesting. I guess I'm confused as to the implications of that when using this in an automated fashion.

I'm a developer on Concourse and our [docker-image-resource](https://github.com/concourse/docker-image-resource) has been expanded to have ECR support by using this library.

We have issues that seem to be caused by the changes in Docker 1.13 and above, which were supposed to be fixed in #26. However, when we bumped our dependency version, it doesn't seem to have fully fixed it. Because of the unique nature of what we do, there's no long-lived stateful `~/.docker/config.json`, and instead we build it on the fly, each time someone uses the resource to build/push an image.

So I'm wondering what the mechanism that causes "No, you need to pull or push an image in the same registry (not repository) or use the registry associated with your account." to make this magically work.

Is it that it populates the particular ECR "auths" section in your `~/.docker/config.json`?

**Comment by samuelkarp** [comment]
There's a cache file for the credentials in `~/.ecr/cache.json`.  When Docker 1.13 and above call `list`, we read this cache file to determine all of the known registries.

https://github.com/docker/docker/pull/32273 is looking to remove the requirement for iterating through all registries during a build and just grab credentials for the specific registries referenced by `FROM` lines.  That's ultimately going to be the best approach here since it means the credential helper won't need to know upfront about some enumerable list.
